src/software/backend/simulation/simulator.cpp

```diff
@@ -1,63 +1,97 @@
-    for (const auto& simulator_robot : friendly_simulator_robots)
+    for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
     {
+        auto simulator_robot    = iter->first;
+        auto& primitive_manager = iter->second;
         SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
-        // TODO: Simulate primitives
-        // app_main(firmware_world.get(), PRIMITIVE STUFF);
+        app_primitive_manager_runCurrentPrimitive(primitive_manager.get(),
+                                                  firmware_world.get());
     }
 
     physics_world.stepSimulation(time_step);
 }
 
 void Simulator::setPrimitives(ConstPrimitiveVectorPtr primitives)
 {
-    // TODO: Set primitives
-    // https://github.com/UBC-Thunderbots/Software/issues/768
+    if (primitives)
+    {
+        auto& primitive_vector = *primitives;
+        for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
+        {
+            auto simulator_robot                        = iter->first;
+            PrimitiveManager* primitive_manager_raw_ptr = iter->second.get();
+
+            auto primitive_iter =
+                std::find_if(primitive_vector.begin(), primitive_vector.end(),
+                             [&simulator_robot](const auto& p) {
+                                 return simulator_robot->getRobotId() == p->getRobotId();
+                             });
+            if (primitive_iter != primitive_vector.end())
+            {
+                // Get params and start primitive
+
+                // The MRFPrimitiveVisitor handles most of the encoding for us
+                MRFPrimitiveVisitor mrf_pv;
+                (*primitive_iter)->accept(mrf_pv);
+                RadioPrimitive_t radio_primitive = mrf_pv.getSerializedRadioPacket();
+                unsigned int primitive_index =
+                    static_cast<unsigned int>(radio_primitive.prim_type);
+                primitive_params_t primitive_params;
+                std::array<double, 4> param_array = radio_primitive.param_array;
+                for (unsigned int i = 0; i < param_array.size(); i++)
+                {
+                    // The data is already scaled appropriately for us from the
+                    // getSerializedRadioPacket function. We just need to pack it
+                    // into an in16_t
+                    double data                = param_array[i];
+                    primitive_params.params[i] = static_cast<int16_t>(std::round(data));
+                }
+
+                primitive_params.slow  = radio_primitive.slow;
+                primitive_params.extra = radio_primitive.extra_bits;
+
+                SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
+                app_primitive_manager_startNewPrimitive(
+                    primitive_manager_raw_ptr, firmware_world.get(), primitive_index,
+                    &primitive_params);
```

Comment:
This will start a new primitive each time, but we probably don't want to do that unless it's actually a new primitive. This was an oversight on my part when I originally built the `PrimitiveManager` abstraction, we should have it encompass a little more, namely this bit of receive.c: https://github.com/UBC-Thunderbots/Software/blob/master/src/firmware/main/io/receive.c#L295

Reply:
will fix in next push (not done yet)

Reply:
done

Reply:
Sorry, I might be missing it, but how are we doing this now? I was expecting some changes to `primitive_manager.{c|h}`

Reply:
whoops, I implemented at the simulator level. If it gets put in the primitive_manager (which it should), that seems to remove the need for separate `start` and `run` primitive functions, otherwise it's implied the caller should know when a primitive gets started. I think it would make sense to just combine them both into a `runPrimitive` function in the primitive manager that handles starting logic internally. Thoughts before I do this?

Reply:
I agree. Thanks for doing this.

Reply:
Updated the primitive_manager, but kept separate `start` and `run` functions. The reason is that these calls happen asynchronously so it's easier to know that one will just start a primitive (if applicable), and one will run it. In the case of the simulator, this saves any confusion or weird cases where when a primitive is started, it gets an extra "tick" of simulation.